### PR TITLE
Fixes #27254 - add support for routes without layout

### PR DESCRIPTION
**Pull request by amirfefer**
This PR adds `RouteLayout` which wraps a component with foreman's layout, therefore pages without this layout (i.e login page) can be rendered with client router, plus plugins would be able to generate their own layout.

### CODE REVIEW

**Review by waldenraines — COMMENTED** [review]
Should we add any tests for this?

File: webpack/assets/javascripts/react_app/routes/LayoutRoute.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import React from 'react';
+import { Route } from 'react-router-dom';
+import PropTypes from 'prop-types';
+import Layout from '../components/Layout';
+
+const RouteLayout = ({ render, layout, ...rest }) => (
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/routes/LayoutRoute.js]: The component name doesn't match the filename.

File: webpack/assets/javascripts/react_app/routes/LayoutRoute.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  <Route
+    {...rest}
+    render={matchProps => <Layout data={layout}>{render(matchProps)}</Layout>}
+  />
+);
+
+RouteLayout.propTypes = {
+  render: PropTypes.func.isRequired,
+  layout: PropTypes.shape({
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/routes/LayoutRoute.js]: Those propTypes appear in the layout component as well? If so, can we reuse?

File: webpack/assets/javascripts/react_app/routes/index.js [hunk 3]
```diff
@@ -1,26 +1,54 @@
...
-          if (railsContainer) railsContainer.remove();
-          currentLocation = renderProps.location;
-
-          return render(renderProps);
-        }}
-      />
-    ))}
-    <Route
+    {routes.map(({ render, path, skipLayout = false, ...routeProps }) => {
+      if (skipLayout) {
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/routes/index.js]: Can this `if` logic move to another component?
It can be simplified to:
```jsx
<ForemanRoute
  beforeRender=(() => removeRootContainer(props))
  {...props}
/>
```

File: webpack/assets/javascripts/react_app/routes/index.js [hunk 4]
```diff
@@ -40,4 +68,32 @@ const AppSwitcher = () => (
   </Switch>
 );
 
+AppSwitcher.propTypes = {
+  data: PropTypes.shape({
+    layout: PropTypes.shape({
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/routes/index.js]: Can be reused as well.

File: webpack/assets/javascripts/react_app/routes/index.js [hunk 5]
```diff
@@ -1,26 +1,54 @@
...
+            return render(renderProps);
+          }}
+          layout={layout}
+        />
+      );
+    })}
+    <LayoutRoute
+      layout={layout}
       render={child => {
         if (
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/routes/index.js:37]: Can this move to a separate function?

File: webpack/assets/javascripts/react_app/ReactApp/ReactApp.js [hunk 6]
```diff
@@ -6,11 +6,9 @@ import history from '../history';
 import Layout from '../components/Layout';
 import AppSwitcher from '../routes';
 
-const ReactApp = ({ data: { layout } }) => (
+const ReactApp = ({ data }) => (
   <Router history={history}>
-    <Layout data={layout}>
-      <AppSwitcher />
-    </Layout>
+    <AppSwitcher data={data} />
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/ReactApp/ReactApp.js:11]: IMO the `data` should not be passed to the children, instead, it should be saved to the redux global state as `app-metadata` so components can use the proper selectors to get this data.

Anyway, I think it should be done on another PR.

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @amirfefer :+1:

**Comment by sharvit** [comment]
> Should we add any tests for this?

+1

**Comment by waldenraines** [comment]
> > Should we add any tests for this?
> 
> +1

@amirfefer waiting for tests here and for other comments to be addressed.

File: webpack/assets/javascripts/react_app/routes/ForemanRoute.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  if (skipLayout) {
+    return (
+      <Route
+        path={path}
+        key={path}
+        {...routeProps}
+        render={renderProps => {
+          beforeRender(renderProps);
+
+          return render(renderProps);
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/routes/ForemanRoute.js:25]: can we rename the `render` prop to `component`? it became
too confusing to have so many `render`s with different meaning. :)

File: webpack/assets/javascripts/react_app/routes/ForemanRoute.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+const ForemanRoute = ({
+  render,
+  layout,
+  skipLayout,
+  beforeRender,
+  path,
+  ...routeProps
+}) => {
+  if (skipLayout) {
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/routes/ForemanRoute.js:16]: In my opinion, instead of having an additional prop,
we can check if the layout doesn't exists, right?
and the default layout could be something simple as: 
```
props => <div>{props.children}</div>`
```

we should also have support for additional layouts.

File: webpack/assets/javascripts/react_app/routes/index.js [hunk 9]
```diff
@@ -1,26 +1,38 @@
...
-      <Route
-        path={path}
+    {routes.map(({ render, path, skipLayout, ...routeProps }) => (
+      <ForemanRoute
         key={path}
+        render={render}
+        layout={layout}
+        path={path}
+        skipLayout={skipLayout}
+        beforeRender={removeRootContainer}
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/routes/index.js:30]: Not sure why is this needed,
is there some bug with rails rendering?

**Review by Ron-Lavi — COMMENTED** [review]
Thanks @amirfefer, Looks great !
left some inline comments :)

File: webpack/assets/javascripts/react_app/routes/ForemanRoute.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        return render(renderProps);
+      }}
+      layout={layout}
+    />
+  );
+};
+
+export default ForemanRoute;
+
+delete layoutPropTypes.history;
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/routes/ForemanRoute.js:46]: This looks a bit strange to me,
if you want to remove some of the layout `proptypes`,
it might be better to use lodash's `omit` method in the layout propTypes definition:
```
layout: PropTypes.shape(omit(layoutPropTypes, 'history')).isRequired,
```

another option is to use patternfly's excludeKeys method on the props.

File: webpack/assets/javascripts/react_app/routes/index.js [hunk 11]
```diff
@@ -1,26 +1,38 @@
...
+const AppSwitcher = ({ data: { layout } }) => (
   <Switch>
-    {routes.map(({ render, path, ...routeProps }) => (
-      <Route
-        path={path}
+    {routes.map(({ render, path, skipLayout, ...routeProps }) => (
+      <ForemanRoute
         key={path}
+        render={render}
+        layout={layout}
```
Comment (xprazak2) [inline webpack/assets/javascripts/react_app/routes/index.js:27]: To have a configurable layout, should each route specify `layout` rather than all routes having the same value from prop?

File: webpack/assets/javascripts/react_app/routes/ForemanRoute.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+import { Route } from 'react-router-dom';
+import PropTypes from 'prop-types';
+import { layoutPropTypes } from '../components/Layout/Layout';
+import RouteWithLayout from './RouteWithLayout';
+import { noop } from '../common/helpers';
+
+const ForemanRoute = ({
+  render,
+  layout,
+  skipLayout,
```
Comment (xprazak2) [inline webpack/assets/javascripts/react_app/routes/ForemanRoute.js:11]: How about passing a falsy value if we do not want a layout? That way you do not need an additional prop:

````js
if (layout) {
// return route with layout
}
// return route without layout
````

**Comment by xprazak2** [comment]
Nice work, but not quite ready yet, needs tests and rebase. Some of the earlier comments are not yet addressed, setting to WoC.

**Comment by amirfefer** [comment]
Thanks @xprazak2, this PR priority is lower than  [adds support for client routing in plugins](https://github.com/theforeman/foreman/pull/6936), which is also a blocker, I will update this PR just after #6936 will be merged :)

**Comment by amirfefer** [comment]
Closing this due to roadmap reprioritizing
